meteor/lib/collections/PartInstances.ts

```diff
@@ -0,0 +1,121 @@
+import * as _ from 'underscore'
+import { TransformedCollection, MongoSelector, FindOptions } from '../typings/meteor'
+import { applyClassToDocument, Time, registerCollection } from '../lib'
+import { Meteor } from 'meteor/meteor'
+import {
+	IBlueprintPartInstance,
+	BlueprintRuntimeArguments,
+	PartEndState,
+} from 'tv-automation-sofie-blueprints-integration'
+import { createMongoCollection } from './lib'
+import { DBPart, Part } from './Parts'
+import { PieceInstance, PieceInstances } from './PieceInstances'
+import { Pieces } from './Pieces'
+
+export interface DBPartInstance extends IBlueprintPartInstance {
+	rundownId: string
+
+	isScratch?: true
+
+	/** Rank of the take that this PartInstance belongs to */
+	takeCount: number
+
+	part: DBPart
+}
+
+export class PartInstance implements DBPartInstance {
+	/** Whether this PartInstance is a temprorary wrapping of a Part */
+	public readonly isTemporary: boolean
+
+	/**
+	 * Whether this PartInstance is a scratch instance - the copy of the Part for the instance
+	 * is still being made and the piece instances are being created.
+	 */
+	public readonly isScratch?: true
+
+	public takeCount: number
+
+	/** Temporarily track whether this PartInstance has been taken, so we can easily find and prune those which are only nexted */
+	public isTaken?: boolean
+
+	// From IBlueprintPartInstance:
+	public part: Part
+	public _id: string
+	public segmentId: string
+	public rundownId: string
+
+	constructor (document: DBPartInstance, isTemporary?: boolean) {
+		_.each(_.keys(document), (key) => {
+			this[key] = document[key]
+		})
+		this.isTemporary = isTemporary === true
+		this.part = new Part(document.part)
+	}
+	getPieceInstances (selector?: MongoSelector<PieceInstance>, options?: FindOptions) {
+		if (this.isTemporary) {
+			throw new Error('Not implemented') // TODO?
+			// const pieces = Pieces.find(
+			// 	{
+			// 		...selector,
+			// 		rundownId: this.rundownId,
+			// 		partId: this._id
+			// 	},
+			// 	{
+			// 		sort: { _rank: 1 },
+			// 		...options
+			// 	}
+			// ).fetch()
+		} else {
+			return PieceInstances.find(
+				{
+					...selector,
+					rundownId: this.rundownId,
+					partInstanceId: this._id
+				},
+				{
+					// sort: { _rank: 1 },
+					...options
+				}
+			).fetch()
+		}
+	}
+	getAllPieceInstances () {
+		return this.getPieceInstances()
+	}
+
+}
+
+export function wrapPartToTemporaryInstance (part: DBPart): PartInstance {
+	return new PartInstance({
+		_id: `${part._id}_tmp_instance`,
+		rundownId: part.rundownId,
+		segmentId: part.segmentId,
+		takeCount: -1, // TODO - is this any good?
+		part: new Part(part)
+	}, true)
+}
+
+export function findPartInstanceOrWrapToTemporary (partInstances: PartInstance[], part: DBPart): PartInstance {
+	return partInstances.find(instance => instance.part._id === part._id) || wrapPartToTemporaryInstance(part)
```

Comment:
Good point, I shall look into those ideas to improve on these.
The performance implications shouldnt be as bad as you fear though, as these should only ever be fed part instances that havent been 'reset' (stopped being loaded. triggered whenever we previously would do a reset of the part ready for the next play of it), but better to be safe and be prepared in case we do want to feed it more.